ambari-server/src/main/java/org/apache/ambari/server/agent/AgentReportsProcessor.java

```diff
@@ -55,40 +58,41 @@ public void addAgentReport(AgentReport agentReport) {
   public AgentReportsProcessor(Configuration configuration) {
 
     ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("agent-report-processor-%d").build();
-    int poolSize = configuration.getAgentsReportThreadPoolSize();
-    executor = Executors.newScheduledThreadPool(poolSize, threadFactory);
-    for (int i=0; i< poolSize; i++) {
-      executor.scheduleAtFixedRate(new AgentReportProcessingTask(),
-          configuration.getAgentsReportProcessingStartTimeout(),
-          configuration.getAgentsReportProcessingPeriod(), TimeUnit.SECONDS);
+    poolSize = configuration.getAgentsReportThreadPoolSize();
+    executors = new ArrayList<>();
+    for (int i = 0; i < poolSize; i++) {
+      executors.add(Executors.newSingleThreadExecutor(threadFactory));
     }
   }
 
   private class AgentReportProcessingTask implements Runnable {
 
+    private final AgentReport agentReport;
+
+    public AgentReportProcessingTask(AgentReport agentReport) {
+      this.agentReport = agentReport;
+    }
+
     @Override
     public void run() {
       try {
         unitOfWork.begin();
-        while (true) {
-          AgentReport agentReport = agentReportsQueue.poll();
-          if (agentReport == null) {
-            break;
-          }
-          String hostName = agentReport.getHostName();
-          try {
-
-            //TODO rewrite with polymorphism usage.
-            if (agentReport.getCommandReports() != null) {
-              hh.handleCommandReportStatus(agentReport.getCommandReports(), hostName);
-            } else if (agentReport.getComponentStatuses() != null) {
-              hh.handleComponentReportStatus(agentReport.getComponentStatuses(), hostName);
-            } else if (agentReport.getHostStatusReport() != null) {
-              hh.handleHostReportStatus(agentReport.getHostStatusReport(), hostName);
-            }
-          } catch (AmbariException e) {
-            LOG.error("Error processing agent reports", e);
+        if (agentReport == null) {
```

Comment:
I don't think `agentReport` can be `null` at this point, since `getHostName().hashCode()` was called earlier on it.